libs/libc/stdio/lib_libvsprintf.c

```diff
@@ -163,6 +163,10 @@ static int vsprintf_internal(FAR struct lib_outstream_s *stream,
   uint16_t flags;
   int width;
   int prec;
+
+  /* For the %m format we may need the current `errno' value */
+
+  int saved_errno = errno;
```

Comment:
This could be used by Kernal threads, right?  Kernel threads do not have errno variables.  Wouldn't this logic just grab garabage out of the Kernel thread's non-existent TLS?

Reply:
even worse, we use syslog even in interrupts.

Reply:
Do you mean through `snprintf` or `syslog`? From what I've seen, both calls are mostly intended for user-space usage and are to be avoided in the kernel, as they're not safe.

Reply:
in nuttx, they are used in kernel as well.

Reply:
> in nuttx, they are used in kernel as well.

syslog is used in assertion and core dumping logic. snprintf is used in various places.

In the PROTECTED and KERNEL build modes, libc can be built without %m by conditioning it on __KERNEL__.  But in the FLAT build mode, there is only one copy of these functions and they must be safe and usable both by application and kernel code.

Reply:
To summarize and take note for the future (feel free to correct me if I'm wrong):

* When building in `protected` and `kernel` modes, the `%m` modifier could be excluded from `libc` by introducing conditional compilation.
* When in flat builds, the code is shared among the user and kernel space.

The problem lies in the fact that the `vsprintf` functions would be using the `errno` variable which points to garbage when calling from the kernel, in any build. Additionally, the code is being shared between user and kernel spaces, making it impossible to differentiate whether the caller is the OS or not.

Potential solutions may imply having kernel-only code for `vsprintf` and relatives, whereas the user space makes use of the POSIX standard functions - this is, the current code. Additionally, some work will be needed to either have a "kernel" `syslog` or to re-route functions that rely on `syslog/vsprintf/...` when in the kernel (ie.: assertion and core dump logic).

Reply:
could we move this statement into `case 'm'` block? @Javinator9889

Reply:
yes, we need carefully review the code to ensure nobody change the errno before enter `case 'm'` block.

Reply:
I've been talking with a teammate here in GEOTAB. One suggestion @patacongo made was to protect that variable when in kernel or protected mode—I assume with an `#ifdef` statement. The problem would be that in flat builds, this will still reference the `errno`, no matter where it is being called from.

However, is it a problem? I mean, when using `vsprintf` from the "kernel" when in flat builds, it should be as simple as avoiding the `%m` modifier, so the `saved_errno` will store garbage that is not used at all - and if a developer decides to use the undocumented `%m` from within the kernel in a flat build, it's up to his responsibility. This will allow user-space to freely use the `%m` modifier.

Opinions? Could this still break something? @xiaoxiang781216 @yamt @patacongo